### DISCUSSION TITLE
get static list of all markdown files

### DISCUSSION THREAD

**Discussion by jcalixte** [ElMassimo/iles#23] [Q&A] [answered]
Hi,

First, congrats for this super cool generator!

I'm trying to replicate what I read from Josh Comeau [here](https://www.joshwcomeau.com/blog/how-i-built-my-blog/#index-pages) with the `getStaticProps` from NextJS and I was wondering how I could get the same result in a îles project.

Thank you 😃

**Comment by ElMassimo** [comment, marked as answer]
__UPDATE__: You can now [`useDocuments('~/posts')`](https://iles-docs.netlify.app/guide/documents#usedocuments).

---

Hi Julien! 😃 

> getLatestContent is a method that traverses the local filesystem to find all of the .mdx blog posts.

This part can be achieved using glob imports, as in the [blog example](https://github.com/ElMassimo/iles/blob/main/playground/the-vue-point/src/logic/posts.ts#L37):

```js
const allPosts = import.meta.globEagerDefault('../pages/posts/**/*.{md,mdx}')
```

To get the __latest__, you can sort them by date. If you use `date` in frontmatter, have in mind that YAML supports [Date objects](https://symfony.com/doc/current/components/yaml/yaml_format.html#dates) so that you don't have to manually do `new Date`.

For the __popular__ part, you could use [`fetch`](https://github.com/developit/unfetch), query any APIs as needed to obtain the names or hrefs of the posts, and then pick the top 5 or sort them accordingly.

<hr/>

Currently îles does not polyfill `fetch` during SSG, but it will in a future release, to simplify this use case.

**Reply by jcalixte** [reply]
Thank you very much, I'll try this!

**Reply by smakinson** [reply]
Been checking out astro a bit and came across îles from a blog post and it looks interesting. Can you confirm if the same limitation exists here (I'm thinking yes since its a Vite feature mentioned), but I thought I'd ask to be sure. Here is the open issue in Astro:

https://github.com/snowpackjs/astro/issues/1700

**Reply by ElMassimo** [reply]
@smakinson Yes, same limitation.

You could use `import.meta.glob` (lazy) to obtain all paths, filter down the paths you are interested in, and await the found module (if any):

Pseudo-code:
```js
async function findPost (slug) {
  const postsByPath = import.meta.glob('./posts/*.mdx')
  return await Object.entries(postsByPath).find(([path, postImportFn]) => {
    // Let's say that the path is enough to know which file to choose:
    if (path.includes(slug)) return postImportFn().then(mod => mod.default)
  })
}
```

Since îles use Suspense, you can easily use this function in a `script setup` without any downsides, or even in `getStaticPaths` and provide the resulting post as props.

By using `glob` instead of `globEager`, it becomes very efficient even if you have a large amount of files, since the modules won't be imported until the import function is called.

The same approach could be used for multiple pages, to narrow down matching files. In the future îles will provide an optimized way to retrieve an entire collection or category.

Do you think it would be useful to create a _Recipes_ section in the docs documenting this kind of approach?

**Reply by smakinson** [reply]
@ElMassimo Thanks for the reply. Yes I believe a recipes section would likely be useful to people as they come in to check it out.
